source/code/plugin/out_mdm.rb

```
@@ -0,0 +1,219 @@
+      end
+      @cached_access_token
+    end 
+
+    def write_status_file(success, message)
+      fn = '/var/opt/microsoft/omsagent/log/MDMIngestion.status'
+      status = '{ "operation": "MDMIngestion", "success": "%s", "message": "%s" }' % [success, message]
+      begin
+        File.open(fn,'w') { |file| file.write(status) }
+      rescue => e
+        @log.debug "Error:'#{e}'"
+        ApplicationInsightsUtility.sendExceptionTelemetry(e.backtrace)
+      end
+    end
+
+    # This method is called when an event reaches to Fluentd.
+    # Convert the event to a raw string.
+    def format(tag, time, record)
+      if record != {}
+        @log.trace "Buffering #{tag}"
+        return [tag, record].to_msgpack
+      else
+        return ""
+      end
+    end
+
+    # This method is called every flush interval. Send the buffer chunk to MDM. 
+    # 'chunk' is a buffer chunk that includes multiple formatted records
+    def write(chunk)
+      if !@first_post_attempt_made || (Time.now > @last_post_attempt_time + retry_mdm_post_wait_minutes*60)
+        post_body = []
+        chunk.msgpack_each {|(tag, record)|
+          post_body.push(record.to_json)
+        }
+        send_to_mdm post_body
+      else
+        @log.info "Last Failed POST attempt to MDM was made #{((Time.now - @last_post_attempt_time)/60).round(1)} min ago. This is less than the current retry threshold of #{@retry_mdm_post_wait_minutes} min. NO-OP"
+      end
+    end
+
+    def send_to_mdm(post_body) 
+      begin
+        access_token = get_access_token
+        uri = URI.parse(@@post_request_url)
+        http = Net::HTTP.new(uri.host, uri.port)
+        http.use_ssl = true
+        request = Net::HTTP::Post.new(uri.request_uri)
+        request['Content-Type'] = "application/x-ndjson"
+        request['Authorization'] = "Bearer #{access_token}"
+        request.body = post_body.join("\n")
+        response = http.request(request)
```

so the magic here is value() raising if non-2xx ? Please ensure that the data indeed go to buffer and get resent next time.

---

Correct. So I simply need to re-raise the exception